### PR TITLE
checker: fixes resolution sum type with an alias type

### PR DESCRIPTION
**Pull request by vincenzopalazzo**
Fixes https://github.com/vlang/v/issues/14029

This PR will introduce the check on the recursive sum type defined with the alias type.

In particular, we are not able to check the following use case, where a `Result<>` is cast to a `ParseRes` because we look only inside the child of the sum type and not if the alias is another subtype.

```
type ParseRes = Result<[]Token, ParseErr>

struct ParseErr{ }

type Opt<T> = None<T> | Some<T>

struct None<T> {}

struct Some<T> {
	value T
}

type Result<T, U> = Err<U> | Ok<T>

struct Ok<T> {
	value T
}

struct Err<U> {
	value U
}

fn test_report() {
	r := Opt<ParseRes>(None<ParseRes>{})
	match r {
		Some<ParseRes> {
			// make possible cast of the same type!
			rx := Result<[]Token, ParseErr>(r.value)
		}
		None<ParseRes> {}
	}
}
```

Hope that the problem is clear, it was very tricky to catch!

Signed-off-by: Vincenzo Palazzo <vincenzopalazzodev@gmail.com>

### CODE REVIEW

**Comment by JalonSolov** [comment]
A lot of JavaScript backend tests are failing...

**Comment by vincenzopalazzo** [comment]
>A lot of JavaScript backend tests are failing...

Right, this is a good, I will add more tests to cover the alias checking, that is pretty tricky!

File: vlib/v/checker/tests/cmp_between_alias_and_sum_type.vv [hunk 1]
```diff
@@ -0,0 +1,10 @@
+struct Empty {}
+type IndexNumber = u32
+type NameIndexType = Empty | IndexNumber
+
+fn main() {
+	elem := NameIndexType(Empty{})
+	if elem is IndexNumber {
+
+	}
+}
```
Comment (spytheman) [inline vlib/v/checker/tests/cmp_between_alias_and_sum_type.vv:10]: Avoid putting code that compiles without errors in `vlib/v/checker/tests/`. Such code is better to be inside `vlib/v/tests/` as part of a normal _test.v file.

Code inside `vlib/v/checker/tests/` is tried to be compiled with -prod, to force warnings to become errors, so that the .out file is stable, which in turn makes successful compilations much longer than they need to be.

Comment (spytheman) [reply vlib/v/checker/tests/cmp_between_alias_and_sum_type.vv:10]: As a bonus, there will be no need for an empty .out file in `vlib/v/tests/`.

File: vlib/v/tests/known_errors/testdata/cast_aliases_on_generics_type.vv [hunk 2]
```diff
@@ -26,7 +24,7 @@ fn test_report() {
 	r := Opt<ParseRes>(None<ParseRes>{})
 	match r {
 		Some<ParseRes> {
-			rx := Result<[]Token, ParseErr>(r)
+			rx := Result<[]Token, ParseErr>(r.value)
 		}
 		None<ParseRes> {}
 	}
```
Comment (spytheman) [inline vlib/v/tests/known_errors/testdata/cast_aliases_on_generics_type.vv:30]: the changes here seem unrelated to the PR ?

Comment (vincenzopalazzo) [reply vlib/v/tests/known_errors/testdata/cast_aliases_on_generics_type.vv:30]: >the changes here seem unrelated to the PR ?

Yes, I originally copy and paste the problem from the issue, but the issue contains the wrong code, r inside the match will begin with type `Some<T>` and not `ParseRes`.

This was confusing to me originally but after a while, I got it with this fix :)

File: vlib/v/checker/tests/cmp_between_alias_and_sum_type.vv (hunk 1 above)
Comment (vincenzopalazzo) [reply vlib/v/checker/tests/cmp_between_alias_and_sum_type.vv:10]: Oh this is good, I was a little bit confused on where to put test that compile with success, thanks for the suggestion

File: vlib/v/tests/cmp_between_alias_and_sum_type.vv [hunk 3]
```diff
@@ -0,0 +1,10 @@
+struct Empty {}
+type IndexNumber = u32
+type NameIndexType = Empty | IndexNumber
+
+fn main() {
+	elem := NameIndexType(Empty{})
+	if elem is IndexNumber {
+
+	}
+}
```
Comment (spytheman) [inline vlib/v/tests/cmp_between_alias_and_sum_type.vv:10]: .vv files are not compiled by default, rename it to `_test.v` .

This is also not a good regression test, because it compiles fine on master too, i.e. it does not demonstrate what the PR fixes.

File: vlib/v/tests/known_errors/testdata/cast_aliases_on_generics_type.vv [hunk 4]
```diff
@@ -1,33 +0,0 @@
...
-
-fn test_report() {
-	r := Opt<ParseRes>(None<ParseRes>{})
-	match r {
-		Some<ParseRes> {
-			rx := Result<[]Token, ParseErr>(r)
-		}
-		None<ParseRes> {}
-	}
-}
```
[the remark this answers is not in the document]
Comment (vincenzopalazzo) [reply vlib/v/tests/known_errors/testdata/cast_aliases_on_generics_type.vv]: You have right, sorry if I removed it!

**Comment by vincenzopalazzo** [comment]
Changed Applied and rebase on master!

Lets wait the CI now 🪑
